### PR TITLE
Overload functions to allow user defined socket.

### PR DESCRIPTION
**Pull request by pardom**
@oliverdain `./gradlew build` fails. Tell me why my changes suck.

### CODE REVIEW

**Comment by pardom** [comment]
Actually, it does build. Still want your review, though.

File: rtmp-client/src/main/cpp/librtmp-jni.c [hunk 1]
```diff
@@ -23,13 +23,14 @@ Java_net_butterflytv_rtmp_1client_RtmpClient_nativeAlloc(JNIEnv *env, jobject in
...
 
+
 /*
  * Class:     net_butterflytv_rtmp_client_RtmpClient
  * Method:    open
  * Signature: (Ljava/lang/String;)I
  */
-JNIEXPORT jint JNICALL Java_net_butterflytv_rtmp_1client_RtmpClient_nativeOpen
-        (JNIEnv * env, jobject thiz, jstring url_, jboolean isPublishMode, jlong rtmpPointer) {
+JNIEXPORT jint JNICALL Java_net_butterflytv_rtmp_1client_RtmpClient_nativeOpenSocket
```
Comment (oliverdain) [inline rtmp-client/src/main/cpp/librtmp-jni.c:32]: It looks like maybe this is auto-generated but if it's not it'd be good to add some comments about what you did here, what socketFd is, etc.

File: rtmp-client/src/main/cpp/librtmp/rtmp.h [hunk 2]
```diff
@@ -314,8 +314,10 @@ extern "C"
 			int dStop, int bLiveStream, long int timeout);
 
   int RTMP_Connect(RTMP *r, RTMPPacket *cp);
+  int RTMP_ConnectSocket(RTMP *r, RTMPPacket *cp, int fd);
```
Comment (oliverdain) [inline rtmp-client/src/main/cpp/librtmp/rtmp.h:317]: Comments here would be helpful. For example:

* You changed this from the existing rtmp so people know why it doesn't match the docs
* The fd == -1 is special
* Why you allow the user to pass in a file descriptor

File: rtmp-client/src/main/cpp/librtmp/rtmp.h [hunk 3]
```diff
@@ -314,8 +314,10 @@ extern "C"
...
   struct sockaddr;
   int RTMP_Connect0(RTMP *r, struct sockaddr *svc);
+  int RTMP_ConnectSocket0(RTMP *r, struct sockaddr *svc, int fd);
```
Comment (oliverdain) [inline rtmp-client/src/main/cpp/librtmp/rtmp.h:320]: as above, comments

File: rtmp-client/src/main/java/net/butterflytv/rtmp_client/RtmpClient.java [hunk 4]
```diff
@@ -49,8 +49,12 @@ public RtmpIOException(int errorCode) {
     }
 
     public void open(String url, boolean isPublishMode) throws RtmpIOException {
+        open(-1, url, isPublishMode);
+    }
+
+    public void open(int socketFd, String url, boolean isPublishMode) throws RtmpIOException {
```
Comment (oliverdain) [inline rtmp-client/src/main/java/net/butterflytv/rtmp_client/RtmpClient.java:55]: comments

File: rtmp-client/src/main/java/net/butterflytv/rtmp_client/RtmpClient.java [hunk 5]
```diff
@@ -61,37 +65,44 @@ public void open(String url, boolean isPublishMode) throws RtmpIOException {
...
+     * @param isPublishMode if this is an publication it is true,
+     *                      if connection is for getting stream it is false
+     * @return return a minus value if it fails
+     * returns
+     * <p>
+     * <p>
      * returns {@link #OPEN_SUCCESS} if it is successful, throws RtmpIOException if it is failed
      */
-    private native int nativeOpen(String url, boolean isPublishMode, long rtmpPointer);
+    private native int nativeOpenSocket(int socketFd, String url, boolean isPublishMode, long rtmpPointer);
```
Comment (oliverdain) [inline rtmp-client/src/main/java/net/butterflytv/rtmp_client/RtmpClient.java:93]: native**Open**Socket is an odd name for a method that doesn't open a socket. Maybe `connectToExistingSocket` or something?

**Review by oliverdain — APPROVED** [review]
Overall looks good. The only real comment is that this needs some documentation. You modified an existing library in some odd ways but didn't document that anywhere. Somebody trying to figure out what you did and why in a month or so would be very confused.

Feel free to merge when you're ready.

**Pull request closed without merging**